File: group-anagrams.js
/*
https://neetcode.io/problems/anagram-groups
Group Anagrams
Given an array of strings strs, group all anagrams together into sublists.
You may return the output in any order.

An anagram is a string that contains the exact same characters as another string,
but the order of the characters can be different.
*/

class Solution {
  /**
   * @param {string[]} strs
   * @return {string[][]}
   */
  groupAnagrams(strs) {

  }

  /**
 * @param {string[]} strs
 * @return {string[][]}
 */
  groupAnagrams2(strs) {
    const res = new Map();

    for (const s of strs) {
      // Create a count array of size 26 (for 'a' to 'z') initialized with 0
      const count = new Array(26).fill(0);

      // Count the occurrences of each character in the string
      for (const c of s) {
        count[c.charCodeAt(0) - "a".charCodeAt(0)]++;
      }

      // Use the count array as a key (converted to a string)
      const key = count.join("#"); // Join with a delimiter to avoid collisions
      console.log('key', key);
      if (res.has(key)) {
        res.get(key).push(s);
      } else {
        res.set(key, [s]);
      }
    }

    // Return all the grouped anagrams
    return Array.from(res.values());
  }
}

// Input: strs = ["act","pots","tops","cat","stop","hat"]
// Output: [["hat"],["act", "cat"],["stop", "pots", "tops"]]

const solution = new Solution()
console.log(solution.groupAnagrams2(["act","pots","tops","cat","stop","hat"]))

File: group-anagrams/group-anagrams.go
package main

import (
	"fmt"
	"sort"
	"strings"
)

/**
https://neetcode.io/problems/anagram-groups
Group Anagrams
Given an array of strings strs, group all anagrams together into sublists.
You may return the output in any order.

An anagram is a string that contains the exact same characters as another string,
but the order of the characters can be different.
**/

func groupAnagrams(strs []string) [][]string {
	var groupedAnagrams [][]string
	var anagramMap = make(map[string][]string)

	for _, word := range strs {
		var sorted = sortString(word)
		var _, exists = anagramMap[sorted]
		if !exists {
			anagramMap[sorted] = []string{word}
		} else {
			anagramMap[sorted] = append(anagramMap[sorted], word)
		}
	}

	for _, value := range anagramMap {
		groupedAnagrams = append(groupedAnagrams, value)
	}

	return groupedAnagrams
}

func groupAnagrams2(strs []string) [][]string {
	letterMap := make(map[[26]int][]string)
	fmt.Println("letterMap", letterMap)

	for _, s := range strs {
		var count [26]int
		for _, c := range s {
			count[c-'a']++
		}
		letterMap[count] = append(letterMap[count], s)
		fmt.Println("count", count, letterMap[count])
	}

	var result [][]string
	for _, group := range letterMap {
		result = append(result, group)
	}
	return result
}

func sortString(w string) string {
	var s = strings.Split(w, "")
	sort.Strings(s)
	return strings.Join(s, "")
}

func main() {
	// var groupedAnagrams = groupAnagrams([]string{"act", "pots", "tops", "cat", "stop", "hat"})
	// fmt.Println("group anagrams", groupedAnagrams, 'c'-'a', 'c', 'a')
	var groupedAnagrams2 = groupAnagrams2([]string{"act", "pots", "tops", "cat", "stop", "hat"})
	fmt.Println("group anagrams", groupedAnagrams2, 'c'-'a')
}
